### ISSUE TITLE
[clang] -Wunused-value does not warn about `literal;` if the literal is expanded from macro defined in system header

### ISSUE THREAD

**Issue by sleeptightAnsiC** [llvm/llvm-project#106867]
Hi,

given following C code:

```C
#include <stdbool.h>
int main (void)
{
    true;
    return 0;
}
```
```
$ gcc -Wunused-value main.c
$ clang -Wunused-value main.c
$ icx -Wunused-value main.c
```

I would expect that `-Wunused-value` would warn about `true;` statement
but it does not - not in GCC 14.2, nor in Clang 18.1, nor in Intel Compiler 2024.1.0
not unlsess I specify `-Wsystem-header`

However,
this functionality was working as expected back in GCC 3.4.6 and Clang 3.0.0
(regression / something has changed? see gobolt link down below)
and recent C++ front-ends DO catch this unlike C front-ends:

```
$ g++ -Wunused-value main.c
main.c: In function ‘int main()’:
main.c:4:9: warning: statement has no effect [-Wunused-value]
    4 |     true;
      |         ^
$ clang++ -Wunused-value main.c
clang++: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]
main.c:4:5: warning: expression result unused [-Wunused-value]
    4 |     true;
      |     ^~~~
```

MSVC 19.40 also catches this with C11 Wall flags:

```
> cl /Wall /nologo /std:c11 main.c
main.c
main.c(4): warning C4555: result of expression not used
```

I've checked everything from GCC bugzilla's meta bug about `-Wunused` warnings and found nothing about it.
I also checked issues in LLVM github repo with the same result - nothing found.

I'm a bit concerned that since so many C compilers give the same result, this might be some kind of "agreed between everyone" behavior, but it looks like a bug so I'm reporting it as a bug :)

If it's not-a-bug, then PLEASE, give some rationale on this topic.
If this an actual bug then I can also send it further to GCC.

References:
https://godbolt.org/z/joK6348rE
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77299
https://stackoverflow.com/questions/78935304/why-wunused-value-does-not-catch-a-statement-true/78935413#78935413

**Comment by AaronBallman** [comment]
It is intentional that we do not diagnose (and it's not just from macros in system headers, but macros in general): https://github.com/llvm/llvm-project/blob/334d1238aafa8ca017d433caaf8f6e00f2622111/clang/lib/Sema/SemaStmt.cpp#L239
which was added in 1c417da558e97f9b77706ed45f90f9c04e6cc085.

Discussion for those changes starts here:
https://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20130114/071975.html
and
https://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20130204/073564.html

I think the reason implementations largely agree is because of tradeoffs for reducing false positives. It's common for us to suppress diagnostics in the presence of macros (and system headers) specifically because of how hard it is to modify those macros to suppress the diagnostic.

https://godbolt.org/z/6jc6x6nPr shows that we disable for system and user macros but not literals.

**Comment by sleeptightAnsiC** [comment]
Thanks for references.

Although I do understand this is probably "as designed" then, just want to point out that clang behavior noticeably differs from gcc here https://godbolt.org/z/xr9Gav88q - gcc does diagnose macros and, with additional `-Wsystem-headers`, also diagnoses the ones from system headers. Clang misses those two cases.

**Comment by AaronBallman** [comment]
Yeah, I'm going to mark this as an enhancement request (rather than a bug); I think there's room for improvement with the behavior here. I think we shouldn't suppress the diagnostic because the expression came from a macro, but should be smarter about when we suppress the diagnostic due to a macro. For example, I think it might make sense to silence the diagnostic when it would appear within a macro definition, but not when it appears due to a macro expansion. e.g.,
```
#define DIAGNOSE_THIS true
#define DIAGNOSE_THIS_TOO (1 + 12)
#define EVEN_DIAGNOSE_THIS(x) (x * 100)

#define DO_NOT_DIAGNOSE_THIS ({ true; 12; })

#define THIS_IS_DEBATABLE ([]{ true; return 12; })
```
I think for `THIS_IS_DEBATABLE`, we should not diagnose `true` as being an unused value (because the user may not be able to control the macro definition) but we should diagnose a use like:
```
void foo() {
  THIS_IS_DEBATABLE; // Diagnose this
  THIS_IS_DEBATABLE(); // This is fine
}
```
